internal/contracts/manager.go

```diff
@@ -877,26 +1022,93 @@ func (cm *contractManager) AddContractListener(ctx context.Context, listener *co
 		} else if len(existing) > 0 {
 			return i18n.NewError(ctx, coremsgs.MsgContractListenerExists)
 		}
+
+		// Check for existense of an older listener with the old signature
+		// Only valid for one filter
+		if listener.Event != nil && len(listener.Filters) == 1 {
+			// Note the event signature has been extended with more information in some blockchain plugins
+			// That is why we do not add the signature in the query but instead iterate over the listeners
+			// and compare the signatures
+			signature := cm.blockchain.GenerateEventSignature(ctx, &listener.Event.FFIEventDefinition)
+			filter := database.ContractListenerQueryFactory.NewFilter(ctx)
+			if existing, _, err := cm.database.GetContractListeners(ctx, cm.namespace, filter.And(
+				filter.Eq("topic", listener.Topic),
+				filter.Eq("location", locationLookup),
+			)); err != nil {
+				return err
+			} else if len(existing) > 0 {
+				for _, listener := range existing {
+					// We have extended the event signature to add more information
+					// So we compare the start with is not guaranteed to be the same
+					// but it's the best comparison
+					if strings.HasPrefix(signature, listener.Signature) {
+						return i18n.NewError(ctx, coremsgs.MsgContractListenerExists)
+					}
+
+				}
+			}
+		}
+
 		return nil
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	if err := cm.validateFFIEvent(ctx, &listener.Event.FFIEventDefinition); err != nil {
+	return &listener.ContractListener, nil
+}
+
+func (cm *contractManager) AddContractListener(ctx context.Context, listener *core.ContractListenerInput) (output *core.ContractListener, err error) {
+	verifiedContractListener, err := cm.verifyContractListener(ctx, listener)
+	if err != nil {
 		return nil, err
 	}
-	if err = cm.blockchain.AddContractListener(ctx, &listener.ContractListener); err != nil {
+
+	if err = cm.blockchain.AddContractListener(ctx, verifiedContractListener); err != nil {
 		return nil, err
 	}
 	if listener.Name == "" {
 		listener.Name = listener.BackendID
 	}
-	if err = cm.database.InsertContractListener(ctx, &listener.ContractListener); err != nil {
+	if err = cm.database.InsertContractListener(ctx, verifiedContractListener); err != nil {
 		return nil, err
 	}
 
-	return &listener.ContractListener, err
+	return verifiedContractListener, err
+}
+
+// Sort the filters by signature then location and generate a hash
+func (cm *contractManager) generateFilterSignature(ctx context.Context, filters core.ListenerFilters) (string, error) {
```

Comment:
Per above this function seems mainly duplicative with the `parseContractListenerFilters` function. Looping through the same structures, and working out the same things.

I do see a `sort` here - which is interesting, whereas in the `parseContractListenerFilters` function we have a map.

Reply:
So we need a sort to make sure that signatures are always the same given any order, it's only duplicative because it generates a signature which I can remove and use the one in the filter instead